source/fundamentals/transactions.txt

```diff
@@ -0,0 +1,400 @@
+database to successfully run the multi-document payment transaction:
+
+A document that represents a customer and associated orders in a collection
+called ``customers``.
+
+.. code-block:: json
+
+   { _id: 98765, orders: [] }
+
+Documents that track inventory items in a collection called ``inventory``.
+
+.. code-block:: json
+
+   [
+    { name: "sunblock", sku: 5432, qty: 85 },
+    { name: "beach towel", sku: 7865, qty: 41 }
+   ]
+
+The code examples also perform operations on the ``orders`` collection, but
+do not require any prior sample documents.
+
+The code examples use the ``cart`` and ``payment`` constants to represent
+a sample list of items purchased and order payment details as follows:
+
+.. code-block:: javascript
+
+   const cart = [
+     { name: 'sunblock', sku: 5432, qty: 1, price: 5.19 },
+     { name: 'beach towel', sku: 7865, qty: 2, price: 15.99 }
+   ];
+
+.. code-block:: javascript
+
+   const payment = { customer: 98765, total: 37.17 };
+
+.. important::
+
+   The examples in the following sections require that you create the
+   collections outside of the transaction or are using MongoDB 4.4 or later.
+   For more information on creating collections inside a transaction, see the
+   :manual:`Create Collections and Indexes in a Transaction </core/transactions/#std-label-transactions-create-collections-indexes>`
+   server guide.
+
+.. _nodejs-transaction-core-api-example:
+
+Core API Implementation
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The code example in this section demonstrates how you can use the Core API to
+run the multi-document payment transaction in a session. This function
+shows how you can perform the following:
+
+1. Start a session
+#. Start a transaction, specifying transaction options
+#. Perform data operations in the same session
+#. Commit a transaction, or cancel it if the driver encounters an error
+#. End a session
+
+.. literalinclude:: /code-snippets/transactions/txn-core.js
+   :language: javascript
+   :emphasize-lines: 2-6,8,10,19,32,41,49,51,55-61,65,67
+   :start-after: start placeOrder
+   :end-before: end placeOrder
+
+Note that you must pass the session object to each CRUD operation that
+you want to run on that session.
+
+The code and comments in ``catch`` block demonstrates how you can identify
+the server transaction errors and where you can place your logic to handle
+them.
+
+See the :ref:`Payment Transaction Data <nodejs-transaction-example-payment-data>`
+to see what your collections should contain after you run the transaction
+successfully.
+
+.. _nodejs-transaction-callback-api-example:
+
+Callback API Implementation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The code examples in this section show how you can use the Callback API to
+run the multi-document payment transaction in a session.
+
+The following code example shows a function you can use to start the
+session and pass the callback function that contains the operations you want
+to run as a transaction.
```

Comment:
[suggestion] I think it would be helpful to include your entire async `run` function here. I thought it was a little confusing to see some of the internal logic of the function without seeing the extra context/ seeing the logic wrapped in a function you can run.

I think if you include the entire function, the only thing missing from a runnable example is an instantiated `MongoClient`

Reply:
Do you mean defining placeOrder() as an anonymous function? If you just meant adding the MongoClient, I don't think that's necessary. We often leave that out of fundamentals pages (e.g. https://docs.mongodb.com/drivers/node/current/fundamentals/crud/write-operations/insert/) because we are helping them learn how to create the transaction rather prescribe the structure or produce a runnable example.

If you meant the former:
What I wanted to show is that you can use a function free of any transaction-specific syntax (unfortunately need the session), but have no idea whether that's an advantage.
I do think inlining would add a lot of clutter since it's not a very short function -- nor would any multi-document transaction -- and therefore draw attention away from the withTransaction parameter list.

Reply:
No, I think placeOrder() is good as-is, in its own code block below. I don't mean adding the MongoClient either, I agree that's not necessary.

I'm suggesting wrapping the snippet that's there in an async function, and maybe including client connection and disconnection. In your code, it's called `run()` and includes consts `cart` and `payment`, as well as the methods to connect and disconnect from the client (you definitely don't need `cart` and `payment` since those are established in the sample data intro. I think connecting/disconnecting could be helpful, but not necessary).

Initially my thinking was to include as much working code as possible to get a user to a working example quickly. Now that I'm reconsidering, I see more value in keeping it as concise as possible to illustrate the topic. Leaving it to you - my thought is that simply wrapping the existing snippet in an async function will add a bit of clarity.

Reply:
Decided not to include connection code. If we find readers request runnable code examples for fundamentals pages, we can address creating a common format then.